K8s

svetomir opened the pull request:
This is large PR which brings two 'big' changes to the beekeeper's design so far.
I could split it in two separate PRs, but I decided not to in order to save time. 
If this is too much for reviewing, please let me know and I will do cherrypicking for 2 PRs.

First 'big' change is related to the **bee** pkg:

- _cluster_ is rewritten so it has more options now
- _nodegroups_ (subset of cluster) are introduced as a concept to allow better management of a cluster and similar bee nodes, they should give more flexibility when creating dynamic clusters
- earlier, cluster use to be []Node, and now, cluster is map[name]NodeGroup, while NodeGroup is map[name]Node
- all this, changes logic how particular Nodes are accessed in the code (checks for example), hence all the checks had to be updated to use new types

Second 'big' change is adding the **k8s** pkg, which has several sub-packages:

- _k8s/bee_ pkg is our implementation of k8s that suits Bee needs, it is a wrapper around other _k8s/*_ pkgs
- other _k8s/*_ pkgs are pure k8s used for CRUD over objects in the k8s cluster

Also, 2 new commands are added:
- start node
- start cluster

acud reviewed (commented):
@svetomir kudos for the tons of work that was put into this PR and topic in general. It seems accessible, well thought through and something we could maintain in the long run.

I did observe some repeating patterns which could be improved. They are generally minor but are apparent in the PR, but also in general in `beekeeper` codebase, and we could probably improve on them slowly but surely.
Namely, those are:
- contexts could be propagated from the cobra command to the actual checks (`cmd.Context()` instead of background context in commands)
- I would personally remove the type aliases in the gist of `type SomeCollection map[string]SomeOtherType`. I feel that in general this lacks clarity of which actual types are being passed around and so I personally find passing around a `map[string]SomeType` as clearer. Also, if the type alias does not provide any OTC functionality that needs this type as a receiver then I would personally see less benefit in having it there
- Some functions return an instance instead of a pointer, example is in the configuration functions. I personally would return a pointer there, since I feel that configuration is something the needs to be passed around with potential mutability in mind
- some stream messages have `2` suffix (not sure why)
- named returns - i would personally avoid them. we also converged on this in the `bee` project that we were abusing them too much. they make some of the flow of the code less readable since you need to look specifically for the actual assignments. i would use named returns only when the actual semantics of the values you're returning is not clear and some labeling is needed (but even then, i am not sure i would use blank `return` statements) or when there is a performance optimization consideration to using them
- some receiver methods use value semantics but i think they should use pointer semantics

Also, there seems to be a lot of repetition with `toK8S` methods, which sometimes require type aliases, and I wonder if we could put something together that does this with reflection, since those type aliases are confusing (sometimes they refer to a slice, sometimes a map), and this makes their usage in general a bit more cumbersome (since you don't necessarily know which underlying plumbing you need to access).

Otherwise this looks good!

svetomir commented on cmd/beekeeper/cmd/create_namespace.go:
```diff
@@ -0,0 +1,44 @@
+				KubeconfigPath: kubeconfig,
+			})
+
+			ctx := context.Background()
```
Good point. Updated it.

svetomir commented on cmd/beekeeper/cmd/print_depths.go:
```diff
@@ -14,30 +14,37 @@ func (c *command) initPrintDepths() *cobra.Command {
+				}
 			}
 
 			ctx := context.Background()
```
Good point. Updated it.

svetomir commented on pkg/k8s/containers/lifecycle.go:
```diff
@@ -0,0 +1,22 @@
+}
+
+// toK8S converts Lifecycle to Kuberntes client object
+func (l Lifecycle) toK8S() *v1.Lifecycle {
```
Good point. Done.

svetomir commented on pkg/k8s/containers/probe.go:
```diff
@@ -0,0 +1,91 @@
+}
+
+// toK8S converts Containers to Kuberntes client object
+func (p Probe) toK8S() *v1.Probe {
```
Good point. Done.

svetomir commented on pkg/k8s/ingress/tls.go:
```diff
@@ -0,0 +1,36 @@
+type TLSs []TLS
+
+// toK8S converts TLSs to Kuberntes client objects
+func (ts TLSs) toK8S() (l []ev1b1.IngressTLS) {
```
Replied in the comment bellow.

svetomir commented on pkg/k8s/ingress/client.go:
```diff
@@ -0,0 +1,57 @@
+}
+
+// Set creates Ingress, if Ingress already exists does nothing
+func (c Client) Set(ctx context.Context, name, namespace string, o Options) (err error) {
```
Replied in the comment bellow.

svetomir commented on pkg/k8s/persistentvolumeclaim/client.go:
```diff
@@ -0,0 +1,57 @@
+}
+
+// Set creates PersistentVolumeClaim, if PersistentVolumeClaim already exists updates in place
+func (c Client) Set(ctx context.Context, name, namespace string, o Options) (err error) {
```
Replied in the comment bellow.

svetomir commented on cmd/beekeeper/cmd/helpers.go:
```diff
@@ -0,0 +1,135 @@
+)
+
+// newBeeDefaultConfig returns default Bee node configuration
+func newBeeDefaultConfig() k8sBee.Config {
```
Good point. I updated it.

svetomir commented on cmd/beekeeper/cmd/helpers.go:
```diff
@@ -0,0 +1,135 @@
+}
+
+// newDefaultNodeGroupOptions returns default node group options
+func newDefaultNodeGroupOptions() bee.NodeGroupOptions {
```
Good point. I updated it.

svetomir commented on cmd/beekeeper/cmd/print_overlay.go:
```diff
@@ -14,30 +14,36 @@ func (c *command) initPrintOverlay() *cobra.Command {
+				}
 			}
 
 			ctx := context.Background()
```
Good point. Updated it.

svetomir commented on cmd/beekeeper/cmd/print_topologies.go:
```diff
@@ -14,36 +14,42 @@ func (c *command) initPrintTopologies() *cobra.Command {
+				}
 			}
 
 			ctx := context.Background()
```
Good point. Updated it.

svetomir commented on cmd/beekeeper/cmd/start_cluster.go:
```diff
@@ -0,0 +1,119 @@
+		Short: "Start Bee cluster",
+		Long:  `Start Bee cluster.`,
+		RunE: func(cmd *cobra.Command, args []string) (err error) {
+			ctx := context.Background()
```
Good point. Updated it.

svetomir commented on cmd/beekeeper/cmd/print_peers.go:
```diff
@@ -14,31 +14,37 @@ func (c *command) initPrintPeers() *cobra.Command {
+				}
 			}
 
 			ctx := context.Background()
```
Good point. Updated it.

svetomir commented on cmd/beekeeper/cmd/start_node.go:
```diff
@@ -0,0 +1,91 @@
+		Short: "Start Bee node",
+		Long:  `Start Bee node.`,
+		RunE: func(cmd *cobra.Command, args []string) (err error) {
+			ctx := context.Background()
```
Good point. Updated it.

svetomir commented on pkg/bee/cluster.go:
```diff
@@ -5,515 +5,293 @@ import (
-		return msgs[i].Index < msgs[j].Index
-	})
+// Overlays returns ClusterOverlays
+func (c *Cluster) Overlays(ctx context.Context) (overlays ClusterOverlays, err error) {
```
Answered bellow

svetomir commented on pkg/bee/cluster.go:
```diff
@@ -5,515 +5,293 @@ import (
-	Error   error
-}
+// ClusterPeers represents peers of all nodes in the cluster
+type ClusterPeers map[string]NodeGroupPeers
```
Answered bellow

svetomir commented on pkg/bee/cluster.go:
```diff
@@ -5,515 +5,293 @@ import (
-		msgs = append(msgs, m)
-	}
+// ClusterSettlements represents settlements of all nodes in the cluster
+type ClusterSettlements map[string]NodeGroupSettlements
```
Answered bellow

svetomir commented on pkg/bee/cluster.go:
```diff
@@ -5,515 +5,293 @@ import (
-	Sent     int
-}
+// ClusterTopologies represents Kademlia topology of all nodes in the cluster
+type ClusterTopologies map[string]NodeGroupTopologies
```
Answered bellow

svetomir commented on pkg/bee/node.go:
```diff
@@ -17,48 +17,56 @@ import (
-	return Node{
-		api: api.NewClient(opts.APIURL, &api.ClientOptions{HTTPClient: &http.Client{Transport: &http.Transport{
+// NewClient returns Bee client
+func NewClient(opts ClientOptions) (c Client) {
```
Updated it.

svetomir commented on pkg/bee/nodegroup.go:
```diff
@@ -0,0 +1,593 @@
+}
+
+// NodeGroupAddresses represents addresses of all nodes in the node group
+type NodeGroupAddresses map[string]Addresses
```
It is :)

svetomir commented on pkg/bee/nodegroup.go:
```diff
@@ -0,0 +1,593 @@
+}
+
+// AddressesStreamMsg2 represents message sent over the AddressStream channel
+type AddressesStreamMsg2 struct {
```
Good find. :) I overlooked it. I had 2 clients at one moment (legacy one, too).

svetomir commented on pkg/check/balances/balances.go:
```diff
@@ -23,18 +24,19 @@ type Options struct {
 // Check executes balances check
-func Check(c bee.Cluster, o Options, pusher *push.Pusher, pushMetrics bool) (err error) {
+func Check(c *bee.Cluster, o Options, pusher *push.Pusher, pushMetrics bool) (err error) {
 	ctx := context.Background()
```
Done.

svetomir commented on pkg/k8s/bee/helpers.go:
```diff
@@ -0,0 +1,328 @@
+	SwarmEnabled        bool
+}
+
+func setInitContainers(o setInitContainersOptions) (inits containers.Containers) {
```
Answered bellow.

svetomir commented on pkg/k8s/bee/node.go:
```diff
@@ -0,0 +1,377 @@
+	portHTTP = 80
+)
+
+// NodeStartOptions ...
```
Added a proper comment.

svetomir commented on pkg/k8s/configmap/configmap.go:
```diff
@@ -0,0 +1,61 @@
+}
+
+// Set creates ConfigMap, if ConfigMap already exists updates in place
+func (c Client) Set(ctx context.Context, name, namespace string, o Options) (err error) {
```
Answered bellow

svetomir commented on pkg/k8s/containers/containers.go:
```diff
@@ -0,0 +1,73 @@
+)
+
+// Containers represents Kubernetes Containers
+type Containers []Container
```
Answered bellow

svetomir commented on pkg/k8s/containers/env.go:
```diff
@@ -0,0 +1,168 @@
+)
+
+// EnvVars represents Kubernetes EnvVars
+type EnvVars []EnvVar
```
Answered bellow

svetomir commented on pkg/k8s/containers/handler.go:
```diff
@@ -0,0 +1,106 @@
+}
+
+// toK8S converts Handler to Kuberntes client object
+func (h Handler) toK8S() v1.Handler {
```
Updated it.

svetomir commented on cmd/beekeeper/cmd/start_cluster.go:
```diff
@@ -0,0 +1,119 @@
+			nConfig := newBeeDefaultConfig()
+			nConfig.Bootnodes = setupBootnodesDNS(bootnodeCount, c.config.GetString(optionNameStartNamespace))
+			for i := 0; i < nodeCount; i++ {
+				if err := ng.StartNode(ctx, bee.StartNodeOptions{
```
Definitely, I will add concurrency, but once I optimise k8s pkg for it. That is next step once this PR is merged.

svetomir commented on cmd/beekeeper/cmd/start_node.go:
```diff
@@ -0,0 +1,91 @@
+		RunE: func(cmd *cobra.Command, args []string) (err error) {
+			ctx := context.Background()
+
+			cluster := bee.NewCluster(clusterName, bee.ClusterOptions{
```
This doesn't seem like a blocker now. I left this when I finish dynamic cluster feature, as I will have more insight on everything needed for it. I planned to refactor this once I deal with priorites.

svetomir commented:
Hey, @acud thanks a lot for the review.  Comments and advices were very detailed and I am grateful for it. I am glad you find it maintainable.
We definitely should improve beekeeper codebase in the long run as now it becomes apparent in which direction and how it should be developed.
I already tackled some of those based on your review. Here are my comments about it:
* contexts could be propagated from the cobra command to the actual checks (cmd.Context() instead of background context in commands)
    * Good point. I updated this.
* I would personally remove the type aliases in the gist of type SomeCollection map[string]SomeOtherType. I feel that in general this lacks clarity of which actual types are being passed around and so I personally find passing around a map[string]SomeType as clearer. Also, if the type alias does not provide any OTC functionality that needs this type as a receiver then I would personally see less benefit in having it there
    * map[string]SomeType, etc. are regular types (not type aliases) needed to mimic k8s types and allow easier type conversion. I used them for clarity because they are mostly part of some more complex type (struct) and this complexity comes with big number of k8s types.
* Some functions return an instance instead of a pointer, example is in the configuration functions. I personally would return a pointer there, since I feel that configuration is something the needs to be passed around with potential mutability in mind
    * Good point. I updated this.
* some stream messages have 2 suffix (not sure why)
    * Good find. It was a legacy from an old client and I overlooked it. Anyways, I renamed it. Thanks :)
* named returns - i would personally avoid them. we also converged on this in the bee project that we were abusing them too much. they make some of the flow of the code less readable since you need to look specifically for the actual assignments. i would use named returns only when the actual semantics of the values you're returning is not clear and some labelling is needed (but even then, i am not sure i would use blank return statements) or when there is a performance optimisation consideration to using them
    * Whole beekeeper codebase is written in this manner and I would keep it as it is. I find it more readable. Anyways, we can review this again with another PR that will be focused on this.
* some receiver methods use value semantics but i think they should use pointer semantics
    * Good point. I updated this.
* Also, there seems to be a lot of repetition with toK8S methods, which sometimes require type aliases, and I wonder if we could put something together that does this with reflection, since those type aliases are confusing (sometimes they refer to a slice, sometimes a map), and this makes their usage in general a bit more cumbersome (since you don't necessarily know which underlying plumbing you need to access).
    * This way we have explicit methods for conversion to k8s types. Also, in the long run we’ll have less troubles replacing them with generics once they become available in Go. :)

janos reviewed (approved):
A lot of great work is here @svetomir. Thank you! My review is a bit subjective, so it is good that @acud gave it a very good one.

acud reviewed (approved):
Thanks for addressing my comments @svetomir. Looks good and thanks again for all the work done here!
